app/src/utils/useDatasets.ts

```diff
@@ -0,0 +1,146 @@
+import { createContext, useContext, useEffect, useState } from "react";
+import DataFetcher from "./DataFetcher";
+import { AllMetadata, Dataset, DatasetStore, LoadStatus } from "./DatasetTypes";
+import Logger from "./Logger";
+
+const METADATA_KEY = "all_metadata";
+const fetcher = new DataFetcher();
+const logger = new Logger(false);
+
+const metadataLoadPromise: Promise<AllMetadata> = fetcher.getMetadata();
+
+interface ResourceCache<R> {
+  resources: Record<string, R>;
+  statuses: Record<string, LoadStatus>;
+}
+
+interface ResourceCacheManager<R> {
+  cache: ResourceCache<R>;
+  setLoadStatus: (resourceId: string, status: LoadStatus) => void;
+  setLoaded: (resourceId: string, resource: R) => void;
+}
+
+/**
+ * Loads a resource and tracks its load state by updating the
+ * ResourceCacheManager.
+ * @param resourceId A unique id to identify the resource. If the same resource
+ *     id is requested multiple times, it will only be fetched once.
+ * @param cacheManager The ResourceCacheManager that tracks that resource.
+ * @param loadFunction A function that loads the resource when called.
+ */
+async function loadResource<R>(
+  resourceId: string,
+  cacheManager: ResourceCacheManager<R>,
+  loadFunction: () => Promise<R>
+) {
+  try {
+    if (cacheManager.cache.statuses[resourceId]) {
```

Comment:
As we talked in slack, I think relying on truthiness of a string feels brittle to me, especially since some of the valid options on the string is "unloaded" and "error". I worry someone may disrupt this functionality with future changes

Reply:
I pulled this out into a helper function shouldLoadResource that returns true if the load status is undefined or "unloaded" or "error". We can adjust that logic later if needed